src/serve-static.ts

```diff
@@ -10,35 +9,82 @@ export type ServeStaticOptions = {
   index?: string // default is 'index.html'
 }
 
-export const serveStatic = (options: ServeStaticOptions = { root: '' }) => {
-  return async (c: Context, next: Next): Promise<Response | undefined> => {
+export const serveStatic = (options: ServeStaticOptions = { root: '' }): MiddlewareHandler => {
+  return async (c, next) => {
     // Do nothing if Response is already set
-    if (c.finalized) {
-      await next()
-    }
+    if (c.finalized) return next()
+
     const url = new URL(c.req.url)
 
     let path = getFilePath({
-      filename: options.path ?? url.pathname,
+      filename: options.path ?? decodeURIComponent(url.pathname),
       root: options.root,
       defaultDocument: options.index ?? 'index.html',
     })
+
     path = `./${path}`
 
     if (existsSync(path)) {
-      const content = readFileSync(path)
-      if (content) {
-        const mimeType = getMimeType(path)
-        if (mimeType) {
-          c.header('Content-Type', mimeType)
-        }
-        // Return Response object
-        return c.body(content)
+      const mimeType = getMimeType(path)
+
+      const stat = lstatSync(path)
+
+      if (mimeType) {
+        c.header('Content-Type', mimeType)
+      }
+
+      c.header('Accept-Ranges', 'bytes')
+
+      c.header('Date', stat.birthtime.toUTCString())
+
+      if (c.req.method == 'HEAD' || c.req.method == 'OPTIONS') {
+        c.header('Content-Length', stat.size.toString())
+
+        c.status(200)
+
+        return c.body(null)
+      }
+
+      const range = c.req.headers.get('range')
+
+      let start = 0
+      let end = stat.size - 1
+
+      if (range) {
+        const parts = range.replace(/bytes=/, '').split('-')
+        start = parseInt(parts[0], 10)
+        end = parts[1] ? parseInt(parts[1], 10) : end
```

Comment:
We have to add the following code if the client range header value exceeds the data size:

```ts
if (stat.size < end - start + 1) {
  end = stat.size - 1
}
```

And should add the tests:

```ts
it('Should return correct headers and data if client range exceeds the data size', async () => {
  const res = await request(server).get('/static/plain.txt').set('range', '0-20')
  expect(res.status).toBe(206)
  expect(res.headers['content-type']).toBe('text/plain; charset=utf-8')
  expect(res.headers['content-length']).toBe('17')
  expect(res.headers['content-range']).toBe('bytes 0-16/17')
  expect(res.text.length).toBe(17)
  expect(res.text).toBe('This is plain.txt')
})
```